### PR TITLE
[WIP] Bug 1960284: use unsupportedConfigOverrides to set prefer-local annotation

### PR DESCRIPTION
**Pull request by candita**
> Change the default ExternalTrafficPolicy from Local to Cluster. In a case where a node that receives traffic doesn’t have a router pod, Cluster permits it to forward to another node with a router pod instead of dropping the traffic.
> 
> Allow the old behavior if the IngressController.Spec.UnsupportedConfigOverrides specifies the ExternalTrafficPolicy as Local.
> 
This PR will be used to implement the prefer-local annotation for unsupportedConfigOverrides.

### CODE REVIEW

**Comment by smarterclayton** [comment]
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws/1390515440211464192
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-azure/1390515492527017984

seems to mitigate most disruption (registry may have an issue)

File: test/e2e/operator_test.go [hunk 1]
```diff
@@ -1920,6 +1920,60 @@ func TestLoadBalancingAlgorithmUnsupportedConfigOverride(t *testing.T) {
 	}
 }
 
+// TestExternalTrafficPolicyConfigOverride verifies that the
+// operator configures router pod replicas to use the "Local" external traffic
```
Comment (Miciah) [inline test/e2e/operator_test.go:1924]:
```suggestion
// operator configures router LoadBalancer service to use the "Local" external traffic
```

**Comment by candita** [comment]
> Hrm, no, it still has disruption. @candita you need to check what the cause of the router disruption is in the linked runs (i.e. what is happening when `disruption_tests: [sig-network-edge] Cluster frontend ingress remain available expand_less` fails, which _might_ be there aren't enough instances, which is the minReady PR i have)

@smarterclayton, @deads2k - I checked in [this log](https://storage.googleapis.com/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.8-upgrade-from-stable-4.7-e2e-aws-ovn-upgrade/1390688879786332160/build-log.txt) and I see an event that describes **"Unreachable -- on openshift.io/frontends-available-test"**.  More details:

`[36mINFO[0m[2021-05-07T18:44:57Z] E0507 18:28:12.303105     264 event_broadcaster.go:253] Server rejected event ..., ReportingController:"openshift.io/frontends-available-test", ReportingInstance:"openshift.io/frontends-available-test-e2e-aws-ovn-upgrade-openshift-e2e-test", Action:"detected", Reason:"Unreachable", Regarding:v1.ObjectReference{Kind:"Route", Namespace:"kube-system", Name:"console", UID:"", APIVersion:"", ResourceVersion:"", FieldPath:""}, Related:(*v1.ObjectReference)(nil), Note:"on new connections", Type:"Warning", DeprecatedSource:v1.EventSource{Component:"", Host:""}, DeprecatedFirstTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeprecatedLastTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeprecatedCount:0}': 'Event "console.167cdc366b15ebd8" is invalid: series.count: Invalid value: "": should be at least 2' (will not retry!) `

**Comment by smarterclayton** [comment]
Note that Azure always preserves source IP when a network level proxy is used, and GCP now supports PROXY in some of their LB modes.

**Comment by candita** [comment]
May 11 02:33:52.083: INFO: cluster upgrade is Failing: Could not update prioritylevelconfiguration "openshift-control-plane-operators" (95 of 695)
...
May 11 02:49:52.087: INFO: cluster upgrade is Failing: Multiple errors are preventing progress:
...
* Could not update clusterrolebinding "helm-chartrepos-view" (452 of 695)
* Could not update configmap "openshift-insights/service-ca-bundle" (496 of 695)
* Could not update deployment "openshift-cluster-node-tuning-operator/cluster-node-tuning-operator" (373 of 695)
* Could not update deployment "openshift-ingress-operator/ingress-operator" (330 of 695)
* Could not update deployment "openshift-operator-lifecycle-manager/catalog-operator" (519 of 695)
* Could not update service "openshift-cluster-storage-operator/cluster-storage-operator-metrics" (426 of 695)
* Could not update service "openshift-cluster-storage-operator/csi-snapshot-webhook" (293 of 695)
* Could not update serviceaccount "openshift-image-registry/node-ca" (310 of 695)
* Could not update servicemonitor "openshift-cluster-samples-operator/cluster-samples-operator" (401 of 695)

/retest

File: pkg/operator/controller/ingress/load_balancer_service.go [hunk 2]
```diff
@@ -400,9 +419,10 @@ func loadBalancerServiceChanged(current, expected *corev1.Service) (bool, *corev
 	updated := current.DeepCopy()
 	changed := false
 
-	// Preserve everything but the AWS LB health check interval annotation &
-	// GCP Global Access internal Load Balancer annotation.
-	// (see <https://bugzilla.redhat.com/show_bug.cgi?id=1908758>).
+	// Preserve everything except:
+	// AWS LB health check interval annotation (see <https://bugzilla.redhat.com/show_bug.cgi?id=1908758>) ,
+	// GCP Global Access internal Load Balancer annotation (see <https://bugzilla.redhat.com/show_bug.cgi?id=1908758>), &
```
Comment (sgreene570) [inline pkg/operator/controller/ingress/load_balancer_service.go:424]: https://bugzilla.redhat.com/show_bug.cgi?id=1908758 isn't applicable to GCP Global Access.
Maybe instead we could use https://issues.redhat.com/browse/NE-518 ?

**Comment by Miciah** [comment]
/retitle [WIP] use unsupportedConfigOverrides to set prefer-local annotation

**Comment by Miciah** [comment]
/retitle [WIP] Bug 1960284: use unsupportedConfigOverrides to set prefer-local annotation

**Comment by Miciah** [comment]
Thanks for your work on this PR!  I borrowed the unsupported config override logic and test code for #622, which supersedes this PR.
/close

**Pull request closed without merging**